### PR TITLE
commander: split protobuf encoding/decoding into new file

### PR DESCRIPTION
**Pull request by benma**
Easier to call from Rust. (work paused again, but putting in this commit so I don't have to look for it again later).

### CODE REVIEW

**Comment by x1ddos** [comment]
Hm... that's a travis script. Doesn't seem to work on github. I'll fix it.

```
+ CONTAINER=shiftcrypto/firmware_v2:16
+ '[' /home/runner/work/bitbox02-firmware/bitbox02-firmware == pull ']'
+ WORKSPACE_DIR=/home/runner/work/bitbox02-firmware/bitbox02-firmware
+ '[' -z /home/runner/work/bitbox02-firmware/bitbox02-firmware ']'
+ TARGET_BRANCH=protobuf
+ '[' '' == true ']'
+ git remote set-branches --add origin protobuf
+ git fetch origin
fatal: couldn't find remote ref refs/heads/protobuf
```

**Comment by x1ddos** [comment]
> Hm... that's a travis script. Doesn't seem to work on github. I'll fix it.
> 
> ```
> + CONTAINER=shiftcrypto/firmware_v2:16
> + '[' /home/runner/work/bitbox02-firmware/bitbox02-firmware == pull ']'
> + WORKSPACE_DIR=/home/runner/work/bitbox02-firmware/bitbox02-firmware
> + '[' -z /home/runner/work/bitbox02-firmware/bitbox02-firmware ']'
> + TARGET_BRANCH=protobuf
> + '[' '' == true ']'
> + git remote set-branches --add origin protobuf
> + git fetch origin
> fatal: couldn't find remote ref refs/heads/protobuf
> ```

Sent https://github.com/digitalbitbox/bitbox02-firmware/pull/499 to fix this.

File: src/commander/protobuf.h [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+#include <compiler_util.h>
+#include <util.h>
+
+#include <hww.pb.h>
+
+#include <stdbool.h>
+
+USE_RESULT bool protobuf_decode(const in_buffer_t* in_buf, Request* request_out);
+
+void protobuf_encode(const Response* response, buffer_t* buf_out);
```
Comment (x1ddos) [inline src/commander/protobuf.h:27]: Doc comments please? Especially for encode which aborts on error but it doesn't say so anywhere.

File: src/commander/protobuf.c [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+#include <pb_decode.h>
+#include <pb_encode.h>
+
+bool protobuf_decode(const in_buffer_t* buf, Request* request_out)
+{
+    pb_istream_t in_stream = pb_istream_from_buffer(buf->data, buf->len);
+    return pb_decode(&in_stream, Request_fields, request_out);
+}
+
+void protobuf_encode(const Response* response, buffer_t* buf_out)
```
Comment (x1ddos) [inline src/commander/protobuf.c:28]: Would've inverted the args position, to match pb_encode and somehow it also makes sense to me in general - the first arg is where the output is. Many stdlib functions like `write(2)` are similar.

**Comment by benma** [comment]
> This is very nice but why stop at commander's protobuf. How about src/backup.c and restore.c?

Unsure what you mean? I can't refactor the whole commander in one PR ;)

The mid term goal is to have Rust use those functions to protobuf encode/decode, and then be able to process api calls in an async function, slowly replacing commander.

> Also, I'm seeing some unused `#include <pb_encode.h>` and `#include <pb_decode.h>` around. 

Where?

**Comment by x1ddos** [comment]
> > This is very nice but why stop at commander's protobuf. How about src/backup.c and restore.c?
> 
> Unsure what you mean?

I meant for backup and restore to also somehow use these new `protobuf_decode` and `protobuf_encode` functions. So, they wouldn't be in src/commander but probably in src/.

>> Also, I'm seeing some unused #include <pb_encode.h> and #include <pb_decode.h> around.
>
> Where?

Like here: https://github.com/digitalbitbox/bitbox02-firmware/blob/2f0ab993/src/restore.c#L26. I don't believe restore uses `pb_encode`. It only decodes. Or did I miss it? Similarly https://github.com/digitalbitbox/bitbox02-firmware/blob/2f0ab993/src/backup.c#L23 and some test/unit-tests - grep for "pb_decode" and "pb_encode".

**Comment by x1ddos** [comment]
Hey, can you rebase onto the latest master? I want to see if github CI is happy now.

**Comment by benma** [comment]
> I meant for backup and restore to also somehow use these new `protobuf_decode` and `protobuf_encode` functions.

But those don't match the signature, the current functions use `Request` and `Response`, and are only for the commander.

unused includes: will check

File: src/commander/protobuf.h (hunk 1 above)
Comment (benma) [reply src/commander/protobuf.h]: done

**Comment by x1ddos** [comment]
> But those don't match the signature, the current functions use `Request` and `Response`, and are only for the commander.

Right. I imagined something like:

```c
// src/protobuf.h
void protobuf_encode(buffer_t* buf_out, const pb_field_t fields[], const void *src);
```

and in Rust I guess you'd have some stronger types.

**Comment by x1ddos** [comment]
Although, now that I think about it, backup/restore use protobufs for a very different purpose. It just happen to be protobufs too like commander, but they don't have to use the same wire format. Probably best to leave backup/restore as is.

**Comment by benma** [comment]
> > But those don't match the signature, the current functions use `Request` and `Response`, and are only for the commander.
> 
> Right. I imagined something like:
> 
> ```c
> // src/protobuf.h
> void protobuf_encode(buffer_t* buf_out, const pb_field_t fields[], const void *src);
> ```
> 
> and in Rust I guess you'd have some stronger types.

nack, can do this in a later step if needed. this seems easier to use in rust, as you can use the exported struct without weird void* casting things.

one step at a time, i don't think backups will be ported anytime soon to rust, so this will go a long way.

**Pull request merged**